Distutils improvements to cmd.pyi and log.pyi

neiljp opened the pull request:
Only two lines from these files were missing for use in some Zulip code, but I've put together most of the public interface for both of these files.

Tests have passed on my github travis already, but I would appreciate any feedback :)

JelleZijlstra commented on stdlib/2and3/distutils/cmd.pyi:
```diff
@@ -13,3 +13,28 @@ class Command:
+    def ensure_dirname(self, option: str) -> None: ...
+
+    def get_command_name(self) -> str: ...
+    def set_undefined_options(self, src_cmd: str, *option_pairs: Sequence[Tuple[str, str]]) -> None: ...
```
This means each entry in `option_pairs` is a sequence of tuples. I haven't checked, but you probably mean `Tuple[str, str]` (each individual arg should be a tuple).

neiljp replied:
The same issue applies in the log.pyi case, I'm sure, ie. `*args` is as below rather than `List[Any]`:
```python
    def log(self, level: int, msg: str, *args: Any) -> None: ...

JelleZijlstra commented on stdlib/2and3/distutils/cmd.pyi:
```diff
@@ -13,3 +13,28 @@ class Command:
+    def get_sub_commands(self) -> List[str]: ...
+
+    def warn(self, msg: str) -> None: ...
+    def execute(self, func: Callable[..., Any], args: List[Any], msg: Optional[str] = ..., level: int = ...) -> None: ...
```
Can the List be a Sequence or even Iterable instead?

neiljp replied:
The only constraint is probably the unpacking, so I would guess `Sequence[Any]`, but I'm not completely sure. The underlying code called is this: (Command.execute calls util.execute)
https://github.com/python/cpython/blob/master/Lib/distutils/util.py#L270

JelleZijlstra replied:
Unpacking works with any Iterable.

JelleZijlstra commented on stdlib/2and3/distutils/log.pyi:
```diff
@@ -0,0 +1,28 @@
+FATAL: int = ...
+
+class Log:
+    def __init__(self, threshold: int=WARN) -> None: ...
```
Defaults should always be `...`.

neiljp replied:
Damn, missed that one! :)

neiljp commented:
This should cover the changes suggested from the review @JelleZijlstra, including using the Iterable and varargs adjustments in log.pyi.

JelleZijlstra commented on stdlib/2and3/distutils/cmd.pyi:
```diff
@@ -13,3 +13,28 @@ class Command:
     @abstractmethod
     def run(self) -> None: ...
+
+    def announce(self, msg: str, level: int = ...) -> None: ...
```
The argument types should probably be Text virtually always; Python 2 users should be able to pass in unicodes.

JelleZijlstra replied:
(So, to be a little more clear, argument types can and should often be Text instead of str, but for return types and a few other contexts, such as arguments to invariant containers like List, it's more complicated.)

neiljp replied:
Right; I think the disconnect is possibly since stubs are python3-style, but are relevant for python2.

JelleZijlstra commented on stdlib/2and3/distutils/cmd.pyi:
```diff
@@ -13,3 +13,28 @@ class Command:
+    def announce(self, msg: str, level: int = ...) -> None: ...
+    def debug_print(self, msg: str) -> None: ...
+
+    def ensure_string(self, option: str, default: Optional[str]) -> None: ...
```
Missing `= ...` for the default.

JelleZijlstra commented on stdlib/2and3/distutils/cmd.pyi:
```diff
@@ -13,3 +13,28 @@ class Command:
+    def debug_print(self, msg: str) -> None: ...
+
+    def ensure_string(self, option: str, default: Optional[str]) -> None: ...
+    def ensure_string_list(self, option: str) -> None: ...
```
This also accepts `List[str]`. (Also, this should actually be `str` and not `Text`; the code does not accept unicode.)

JelleZijlstra commented on stdlib/2and3/distutils/cmd.pyi:
```diff
@@ -13,3 +13,28 @@ class Command:
+    def warn(self, msg: str) -> None: ...
+    def execute(self, func: Callable[..., Any], args: Iterable[Any], msg: Optional[str] = ..., level: int = ...) -> None: ...
+    def mkpath(self, name: str, mode: int = ...) -> None: ...
+    def copy_file(self, infile: str, outfile: str, preserve_mode: int = ..., preserve_times: int = ..., link: Optional[str]= ..., level: Any = ...) -> Tuple[str, bool]: ...  # level is not used
```
missing a space before = (we should really teach the linter about that)

JelleZijlstra commented on stdlib/2and3/distutils/cmd.pyi:
```diff
@@ -13,3 +13,28 @@ class Command:
+    def copy_file(self, infile: str, outfile: str, preserve_mode: int = ..., preserve_times: int = ..., link: Optional[str]= ..., level: Any = ...) -> Tuple[str, bool]: ...  # level is not used
+    def copy_tree(self, infile: str, outfile: str, preserve_mode: int = ..., preserve_times: int = ..., preserve_symlinks: int = ..., level: Any = ...) -> List[str]: ...  # level is not used
+    def move_file(self, src: str, dest: str, level: Any = ...) -> str: ...  # level is not used
+    def spawn(self, cmd: List[str], search_path: int = ..., level: Any = ...) -> None: ...  # level is not used
```
The cmd can be a `Sequence[Text]` or even `Iterable[Text]` I think. https://github.com/python/cpython/blob/master/Lib/distutils/spawn.py says it's documented as list, but the code actually supports any iterable.

neiljp replied:
Agreed, via the `list(cmd)`.

JelleZijlstra commented on stdlib/2and3/distutils/cmd.pyi:
```diff
@@ -13,3 +13,28 @@ class Command:
+    def move_file(self, src: str, dest: str, level: Any = ...) -> str: ...  # level is not used
+    def spawn(self, cmd: List[str], search_path: int = ..., level: Any = ...) -> None: ...  # level is not used
+    def make_archive(self, base_name: str, format: str, root_dir: Optional[str] = ..., base_dir: Optional[str] = ..., owner: Optional[str] = ..., group: Optional[str] = ...) -> str: ...
+    def make_file(self, infiles: Union[str, List[str], Tuple[str]], outfile: str, func: Callable[..., Any], args: List[Any], exec_msg: Optional[str] = ..., skip_msg: Optional[str] = ..., level: Any = ...) -> None: ...  # level is not used
```
Can infiles be a more general type like `Iterable[str]`? Haven't traced the code to check.

neiljp replied:
The function raises a `TypeError` if not those types:
https://github.com/python/cpython/blob/ede2ac913eba47131ee1bbc37a9aea344d678576/Lib/distutils/cmd.py#L386

JelleZijlstra commented on stdlib/2and3/distutils/log.pyi:
```diff
@@ -0,0 +1,28 @@
+from typing import List, Any, Callable, Iterable
+
+DEBUG: int = ...
```
You don't need the `= ...` part.

JelleZijlstra commented on stdlib/2and3/distutils/log.pyi:
```diff
@@ -0,0 +1,28 @@
+LogFunc = Callable[[str, Iterable[Any]], None]
+
+log: Callable[[int, str, Iterable[Any]], None]
+debug: LogFunc = ...
```
Same here: you don't need `...`

JelleZijlstra commented on stdlib/2and3/distutils/log.pyi:
```diff
@@ -0,0 +1,28 @@
+
+LogFunc = Callable[[str, Iterable[Any]], None]
+
+log: Callable[[int, str, Iterable[Any]], None]
```
Why isn't this just `log: LogFunc`?

neiljp replied:
`log` is the more general form, supporting the extra `level: int` parameter.

JelleZijlstra commented on stdlib/2and3/distutils/log.pyi:
```diff
@@ -0,0 +1,28 @@
+    def error(self, msg: str, *args: Any) -> None: ...
+    def fatal(self, msg: str, *args: Any) -> None: ...
+
+LogFunc = Callable[[str, Iterable[Any]], None]
```
This should be private (start with _) to make clear that it's not an exported name in the real module.

neiljp commented:
Rebased & updated based on above feedback.

I've changed some `str` -> `Text`, but would appreciate feedback on those changes and whether others are good to do too - eg. for filenames?

JelleZijlstra commented:
Thanks! By the way, we usually prefer people don't amend their commits, because it dissociates previous review comments from the code and makes it harder to verify that requested changes have been made. Just make your changes in a new commit and push that.
